Multiple env files (and other improvements)

hcs42 opened the pull request.

hcs42 committed:
fix typos

hcs42 committed:
Use precise match when finding the project name

Previously, the "docker compose ls ... Name={project_name}" filter found
projects whose name contained {project_name} as a substring. For
example, if the project name was "my_project", it also found
"this_is_my_project".

The Docker documentation mentions that "You can also filter for a
substring in a name"
(https://docs.docker.com/reference/cli/docker/container/ls/), but for
some reason it's not documented that the specified "name" is actually
used as a regular expression
(https://github.com/docker/docs/issues/11767).

This commit modifies dkp to find (and archive) only the specified
project.

According to
https://docs.docker.com/compose/project-name/#set-a-project-name,
"project names must contain only lowercase letters, decimal digits,
dashes, and underscores", so we can safely use the project name as a
regular expression.

hcs42 committed:
Support multiple env files

Previously dkp assumed that:

*   The .env and *.env files next to the first compose file were the env
    files shall be copied into the archive.

*   The "docker compose" commands don't require "--env-file <file>"
    options to work.

After this commit, the users can specify a list of env files with the
option "--env-file <file1> <file2>...".

*   These (and only these) env files are copied into the archive.

*   These env files are passed to the "docker compose" commands.

If no "--env-file" option is specified, dkp works the same way as
before.

hcs42 committed:
Fix copying multiple env files

Previously the env file copying worked properly only when there was one
env file and it was called ".env".

*   If there were multiple env files, including .env, dkp copied the
    content of .env into all of them (because "info.env_file" always
    pointed to ".env").

*   If there were multiple env files, but no ".env", dkp crashed:

    FileNotFoundError: [Errno 2] No such file or directory:
    '.../test-data/.env'

Now dkp copies every .env and *.env file with its own content.

hcs42 committed:
Fix: Pass project name to "docker compose"

Previously dkp did not export the volumes of the project if the
project's directory was not the same as the project's name. This was
because "docker compose config" assumed that the project's name is the
directory, and created the volume names based on that.

For example, if I have a directory called "my-dir", but I always call
"docker compose" with "-p my-project" (or I set
COMPOSE_PROJECT_NAME=my-project), and I have volume called "my-volume",
then the volume's real name will be "my-project_my-volume". However,
when I issued "dkp my-project" (without setting COMPOSE_PROJECT_NAME),
it didn't pass "-p my-project" to docker compose, so docker compose
printed a config that used "my-dir_my-volume" as the volume name, and
dkp tried to export that volume.

After this commit, we pass the project name to docker compose, so it
will print the volumes with the correct names, so dkp can export them.

hcs42 committed:
Export non-prebuilt images

Let's say that we use the following compose file for the test-data
project:

    services:
      dummy:
        image: busybox
      web:
        build:
          context: .
          dockerfile_inline: |
            FROM nginx:alpine

When the user starts the project, Compose builds and stores the nginx
image as "test-data_web:latest".

Before this commit, dkp exported only the busybox image.

After this commit, the default behavior stays the same, but if the user
passes the "--all-images" option to dkp, it will also export the
"test-data_web:latest" image.

hcs42 committed:
Fix typing problems

Before:

    $ mypy dkp.py
    dkp.py:226: error: Need type annotation for "args" (hint: "args: List[<type>] = ...")  [var-annotated]
    dkp.py:359: error: Incompatible types in assignment (expression has type "Path", variable has type "str")  [assignment]
    dkp.py:364: error: Unsupported left operand type for / ("str")  [operator]
    dkp.py:372: error: Unsupported left operand type for / ("str")  [operator]
    dkp.py:377: error: Unsupported left operand type for / ("str")  [operator]
    dkp.py:423: error: Argument 1 to "gen_scripts" has incompatible type "str"; expected "Path"  [arg-type]
    dkp.py:431: error: Argument 3 to "encrypt" has incompatible type "Optional[str]"; expected "str"  [arg-type]
    Found 7 errors in 1 file (checked 1 source file)

After:

    $ mypy dkp.py
    Success: no issues found in 1 source file

reddec commented:
Looks interesting. Could you please describe a little bit what are you going to introduce?

hcs42 commented:
The general theme of this PR is that previously `dkp` worked well when a project used Docker and Docker Compose with the default configuration, but it didn't support some custom configuration that we (the company where I work) had in our projects.

For example, `dkp` worked well when the compose project name was same as the directory, and it worked well when the only env file was the `.env` file. This PR makes `dkp` work when the compose project name is different from the directory name, and when we have multiple env files with various names (not necessarily `*.env`).

The first four commits are bugfixes:

* Commit 22f6fc4 fixes typos.
* Commit 60bbe33 makes `dkp` work well when there are multiple projects with similar names (e.g., `my_project` and `this_is_my_project`).
* Commit a872aa3 makes `dkp` copy multiple env files correctly.
* Commit f9c24e3 makes `dkp` work well when the project name and directory name differs.

The next two commits are features:

* Commit 24fbd86 adds support for multiple env files.
    - Our projects often use multiple env files, and the names of these files don't match `*.env`. We pass the list of env files to Docker Compose with the `--env-file` option every time we call a Docker Compose command. So it seemed to make sense to introduce the same option to `dkp`.
* Commit cdbba85 adds support for exporting non-prebuilt images.
    - The motivation here was that:
        1. We wanted to transfer Compose packs to machines without Internet.
        2. My compose file contained images built by the compose file itself (as opposed to having an `image` key in the compose file referring to an existing image). See an equivalent example in the commit message. I called these images non-prebuilt images.
    - Without this commit, `dkp` created a pack that does *not* contain the non-prebuilt images. So when I transferred the pack to the machine without Internet, it could not create the containers because it first wanted to pull down the base images from Dockerhub (but could not).
    - It seemed logical to extend `dkp` with an option (I called it `--all-images`) that asks `dkp` to add even these images into the pack. Maybe we don't even need an option, and this is how `dkp` should always behave. But I wasn't sure about the use cases of other people, so my commit doesn't change the default behaviour.

The last commit is about typing:

- Commit 208a5a0 fixes issues reported by the "mypy" type checker. It would have been nicer to make this the first commit, but I realized that I should run mypy only when I completed all the other commits, so I created this commit on the top of the others.

You can find even more detail in the commit messages.

reddec commented on dkp/dkp.py:
```diff
@@ -267,22 +304,42 @@ def gen_scripts(
     project_name: str,
     output: Path,
-    password: Union[str, None] = None,
+    password: str,
```
Password can be None if not set

hcs42 replied:
> Password can be None if not set

Currently dkp doesn't work without a password:

```
$ ../dkp/dkp.py -o test.pack test-data
Traceback (most recent call last):
  File "/home/hcs/r/dkp/test-data/../dkp/dkp.py", line 521, in <module>
    main()
  File "/home/hcs/r/dkp/test-data/../dkp/dkp.py", line 506, in main
    args.passphrase is not None and args.passphrase != ""
                                    ^^^^^^^^^^^^^^^^^^^^^
AssertionError: Passphrase is not set via argument neither via PASSPHRASE environment variable
```

If the intention is to work without a password (which would be nice actually), we would need to modify the `backup` function not to call `encrypt` when `password` is `None`:

```diff
-       print("encrypting...")
-       encrypted_file = output.with_name(archive_file.name + ".gpg")
-       encrypt(archive_file, encrypted_file, password)
-       # replace archive
-       fs_move(encrypted_file, archive_file)
+       if password is not None:
+           print("encrypting...")
+           encrypted_file = output.with_name(archive_file.name + ".gpg")
+           encrypt(archive_file, encrypted_file, password)
+           # replace archive
+           fs_move(encrypted_file, archive_file)
```

Shall I implement this? I wouldn't change my existing commit (because that is in-line with the fact that currently dkp makes password mandatory). I would to it in a separate commit.

reddec replied:
Thanks! Yes, it supposed to work without password 😅

hcs42 replied:
I pushed a commit about making the passphrase optional: https://github.com/reddec/dkp/pull/2/commits/81c5c86c300a5dcce7f5ab4fe95d26ea543bcfc1

reddec commented on dkp/dkp.py:
```diff
@@ -267,22 +304,42 @@ def gen_scripts(
     make_executable(restore)
 
 
+def copy_env_file(source: Path, target: Path, work_dir: Path):
```
nit: there is nothing specific for `env` files in this function. Plain `copy_file` might be more universal name.

The `work_dir` is practically not used except for logging. I think the argument can be removed to avoid confusion.

hcs42 replied:
Good point. My new commit eliminates this function altogether.

reddec commented on dkp/dkp.py:
```diff
@@ -346,21 +403,21 @@ def backup(
-                env_file.relative_to(work_dir),
-            )
-            fs_copy(info.env_file, env_file)
+        if env_files == []:
+            file: Path
+            for file in info.work_dir.iterdir():
+                if (
+                    file.name != ".env" and file.suffix != ".env"
+                ) or file.resolve().is_dir():
+                    continue
+                copy_env_file(file, project_dir / file.name, Path(work_dir))
+        else:
+            env_file: Path
+            for env_file in env_files:
+                copy_env_file(
+                    env_file,
+                    project_dir / env_file.name,
+                    Path(work_dir))
```
I afraid it means that `--env-file` and default behaviour are mutually exclusive. Shall we consider combining them together?

My suggestion is to iterate over all env_files (as in second branch) plus on top iterate over default env files in work dir (as in first branch).

But I am open for discussion. If there are strong reasons to keep totally different behaviour, then let's just reflect it in documentation.

hcs42 replied:
> My suggestion is to iterate over all env_files (as in second branch) plus on top iterate over default env files in work dir (as in first branch).

I started along this line, with the exception that when at least one env file is specified via `--env-file`, there is no need to copy `.env`. (If you specify at least one `--env-file` to `docker-compose`, it won't use your `.env` file.)

Then I realized that my implementation doesn't handle two cases well.

1.  It doesn't handle it well when `--env-file` is passed, and there are also `*.env` files used by the `env_file` attribute of the compose file. This is related to your remark.

2.  dkp so far copied all env files into the target root directory (`work_dir`) inside the package. This was OK before I added the `--env-file` option because all files were copied from the project root directory (`info.work_dir`). In short: `cp <project_dir>/<env-file-name> <target_project_dir>/<env-file-name>` worked.

    But `--env-file` made it possible to specify files at other locations, not in the root directory of the project. These cases were not handled well in my commit. Here is how we should handle these cases:

    * (a) The env file is outside the project directory.
        - Example: `docker compose --env-file ../my.env`
        - In this case we should skip the env file (just like dkp skips bindigs outside the project directory).
    * (b) The env file is inside a subdirectory in the project directory (maybe multiple directories deep).
        - Example: `docker compose --env-file dir/my.env`
        - In this case we should create the subdirectory in the pack, and copy the env file there.

I modified the code to handle these cases correctly: https://github.com/reddec/dkp/pull/2/commits/fbc2b02f6412daedf5251b0ac47aca6261126b89

It's still not perfect. I documented a remaining limitation in the commit message.

reddec commented:
Overall, @hcs42 thanks a lot for the PR! It is a very useful and nice addition. Just few polishing left.

hcs42 committed:
Improve copying env files

1.  Previously when the user passed the `--env-file` option to dkp, and
    the compose file used `*.env` files with the `env_file` attribute,
    dkp did not copy these `*.env` files.

    Now dkp copies the `*.env` files in this case.

    This is still not perfect because dkp's heuristic is that it copies
    all `*.env` files that it finds in the project root directory.
    Sometimes a file is copied unnecessarily (when it is not used), and
    sometimes a file is not copied that should be (when an env file does
    not have a *.env filename, or it is located in a subdirectory).

    A more comprehensive solution would be to collect the env file names
    from the `env_file` attributes of the composed files.

2.  Previously dkp copied all env files into the target root directory
    (`work_dir`). This was fine for those env files are are in the
    project root directory, but this was not fine for other env files.

    Now:

    *   dkp skips the env files that are outside the project directory.
        dkp could not restore this files on the target machine, so the
        user should take care about them. That's why now dkp prints
        messages about skipping these files.

    *   dkp re-creates the directory structure for those env files that
        are in subdirectories inside the project directory.

hcs42 committed:
Make passphrase optional

Previously specifying a passphrase for the pack was mandatory. This
could be done either by using the "--passphrase"/"-p" option, or by
using the PASSPHRASE environment variable.

This commit makes the passphrase optional. This way the users can create
unencrypted packages, and they don't need GPG to be installed.

reddec commented on dkp/dkp.py:
```diff
@@ -267,12 +304,16 @@ def gen_scripts(
         nice_args = " " + src_args
+
+    for env_file in info.env_files:
+        nice_args += f" --env-file '{env_file}'"
```
Might be strange corner case with env_file with `'` in name

reddec replied:
AFAIK there is something like https://docs.python.org/3/library/shlex.html#shlex.quote

reddec commented:
Thanks!

reddec commented:
@hcs42 I afraid something wrong with tests. Did you try `make test`?

hcs42 commented:
> @hcs42 I afraid something wrong with tests. Did you try `make test`?

Oh, I didn't realize that there were automatic tests... I will look into it.

hcs42 commented:
Does `make test` pass for you locally on v0.2.0?

I get the following output:

```
$ git checkout v0.2.0
HEAD is now at 4954f4f release v0.2.0

$ make test
cat .github/workflows/release.yaml | yq '.jobs.test.steps[].run // "" ' | bash --noprofile --norc -eo pipefail
bash: line 1: cd test-data\ndocker compose -f docker-compose.yaml -f docker-compose.2.yaml create --pull missing\ndocker compose -f docker-compose.yaml -f docker-compose.2.yaml up -d\ncd ../\necho "waiting for start"\nsleep 5\n\n# add data to shared volume\ncd test-data\necho "Payload" | docker compose -f docker-compose.yaml -f docker-compose.2.yaml exec -T dummy /bin/sh -c 'cat - > /data/helloworld'\ncd ../\n: No such file or directory
make: *** [Makefile:5: test] Error 127
```

I have installed the newest versions of poetry and yq:

```
$ pip list | grep poetry
poetry               1.8.2
poetry-core          1.9.0
poetry-plugin-export 1.7.1

$ pip list | grep yq
yq                   3.4.3
```

When I run the first two commands in the `make test` pipeline, I got the following output:

```
$ cat .github/workflows/release.yaml | yq '.jobs.test.steps[].run // "" '
"cd test-data\ndocker compose -f docker-compose.yaml -f docker-compose.2.yaml create --pull missing\ndocker compose -f docker-compose.yaml -f docker-compose.2.yaml up -d\ncd ../\necho \"waiting for start\"\nsleep 5\n\n# add data to shared volume\ncd test-data\necho \"Payload\" | docker compose -f docker-compose.yaml -f docker-compose.2.yaml exec -T dummy /bin/sh -c 'cat - > /data/helloworld'\ncd ../\n"
"data=\"$(curl -v http://127.0.0.1:8080/helloworld)\"\n[ \"$data\" = \"Payload\" ]\n"
"rm -rf backups\n./dkp/dkp.py -o backups/output.bin -p foobar test-data\n"
"docker stop $(docker ps -q)\ndocker rm $(docker ps -aq)\ndocker volume rm $(docker volume ls -q)\ndocker rmi $(docker images -aq)\n"
"cd backups\n./output.bin -r -s foobar\ncd ..\necho \"waiting...\"\nsleep 5\n"
"data=\"$(curl -v http://127.0.0.1:8080/helloworld)\"\n[ \"$data\" = \"Payload\" ]\n"
```

The problem is that Bash tries to interpret the first line as one command, instead of a series of commands.

In the meantime, I will try to fix the tests by running them on GitHub in my fork.

hcs42 commented:
I found the reason of the test failure when running on GitHub, and fixed it in PR #3.
